File: ilp/HM_ilp.cpp
#include "gurobi_c++.h"
using namespace std;

#include <fstream>       // std::ifstream
#include <iostream>      // cin, cout
#include <map>           //map
#include <string>        // string
#include <unordered_map> //unordered_map
#include <unordered_set> // unordered_set
#include <utility>       // pair
#include <vector>        // vector

int k = 0;
int tau = 0;

// Data structure for all the information needed
struct Input {
  unordered_map<string, int> kmers_frequency;
  unordered_set<string> forbiden_patterns;
  vector<char> alphabet;
  // map instead of unordered map because hashes of pair are not defined
  map<pair<string, string>, int> context_hashmark_index;
  vector<int> context_hashmark_count;
  int nb_hashmark;
  int nb_critical;
  map<string, int> criticals_index;
  vector<string> list_criticals;
  vector<vector<vector<int>>> alpha;
  vector<pair<int, int>> forbiden_replacements;
};

bool is_critical(string kmer, Input &input) {
  return (input.kmers_frequency[kmer] < tau &&
          input.kmers_frequency[kmer] + k * input.nb_hashmark >= tau);
}

void count_critical(int i_, int j_, string replaced, Input &input) {
  for (int it = 0; it <= replaced.size() - k; it++) {
    std::string kmer = replaced.substr(it, k);
    if (is_critical(kmer, input)) {
      if (input.forbiden_patterns.count(kmer) == 1) {
        input.forbiden_replacements.push_back(make_pair(i_, j_));
      } else if (input.criticals_index.count(kmer) == 0) {
        // First time we see this critical kmer
        input.criticals_index[kmer] = input.nb_critical;
        input.list_criticals.push_back(kmer);
        vector<vector<int>> coeff(input.context_hashmark_index.size(),
                                  vector<int>(input.alphabet.size(), 0));
        input.alpha.push_back(coeff);
        input.nb_critical++;
      }
      // account for the kmer added
      input.alpha[input.criticals_index[kmer]][i_][j_]++;
    }
  }
}

// Parse the input and get all informtion needed, occurences of kmer, context
// etc
Input parse_input(string input_file, string forbiden_pattern_file) {
  Input input;
  vector<string> hashmark;
  unordered_set<char> alphabet;
  ifstream is(input_file); // open file
  char c;
  string window = "";
  bool hash_has_back_context = true;
  while (is.get(c)) {
    if (c == '\n') {
      continue;
    }
    if (c == '#') {
      hashmark.push_back(window + "#");
      window = "";
      hash_has_back_context = false;
    } else {
      alphabet.insert(c);
      if (window.size() < k - 1)
        window += c;
      else {
        if (!hash_has_back_context) {
          hashmark.back() += window;
          hash_has_back_context = true;
        }
        if (input.kmers_frequency.count(window) == 0)
          input.kmers_frequency[window] = 0;
        input.kmers_frequency[window]++;
        window += c;        // add new character
        window.erase(0, 1); // remove first character
      }
    }
  }
  is.close(); // close file
  input.nb_hashmark = hashmark.size();
  // Parse forbiden_patterns
  ifstream is_forbiden_patterns(forbiden_pattern_file); // open file
  string pattern;
  while (getline(is_forbiden_patterns, pattern))
    input.forbiden_patterns.insert(pattern);
  input.alphabet.assign(alphabet.begin(), alphabet.end());

  int hashmark_index = 0;
  input.nb_critical = 0;

  // convert hashmark in unique context, count them
  for (string &hash : hashmark) {
    auto context = make_pair(hash.substr(0, k - 1), hash.substr(k, k - 1));
    if (input.context_hashmark_index.count(context) == 0) {
      input.context_hashmark_index[context] = hashmark_index;
      input.context_hashmark_count.push_back(0);
      hashmark_index++;
    }
    input.context_hashmark_count[input.context_hashmark_index[context]]++;
  }

  // Compute the critical kmer and the number we may add through replacement
  // (alpha)
  for (map<pair<string, string>, int>::iterator it =
           input.context_hashmark_index.begin();
       it != input.context_hashmark_index.end(); ++it) {
    auto context = it->first;
    for (int l = 0; l < input.alphabet.size(); l++) {
      string replaced = context.first + input.alphabet[l] + context.second;
      count_critical(it->second, l, replaced, input);
    }
  }

  return input;
}

int main(int argc, char *argv[]) {
  if (argc != 5)
    cout << "You must give 4 arguments: "
         << "k tau input_file forbiden_patterns" << endl;
  k = stoi(argv[1]);
  tau = stoi(argv[2]);
  string input_file = argv[3];
  string forbiden_pattern_file = argv[4];

  Input input = parse_input(input_file, forbiden_pattern_file);
  cout << "Finished parsing input to get stats!" << endl;

  try {

    // Create an environment
    GRBEnv env = GRBEnv(true);
    env.set("LogFile", "HM_ilp.log");
    env.start();

    // Create an empty model
    GRBModel model = GRBModel(env);

    // Create variables
    int nb_context = input.context_hashmark_index.size();
    GRBVar **x = new GRBVar *[nb_context];
    for (int i = 0; i < nb_context; i++)
      x[i] = model.addVars(input.alphabet.size(), GRB_INTEGER);

    GRBVar *z = model.addVars(input.nb_critical, GRB_BINARY);

    // Create objective
    GRBLinExpr obj = 0;
    for (int l = 0; l < input.nb_critical; l++)
      obj += z[l];
    // Set objective: minimize the sum of z
    model.setObjective(obj, GRB_MINIMIZE);

    // x[i][j] non negatives
    for (int i = 0; i < nb_context; i++) {
      for (int j = 0; j < input.alphabet.size(); j++) {
        model.addConstr(x[i][j] >= 0, "non negative x");
      }
    }

    // No forbiden replacement
    for (auto &r : input.forbiden_replacements) {
      model.addConstr(x[r.first][r.second] == 0, "no sensitive pattern");
    }

    GRBLinExpr lhs;
    // limit added kmer or consider it as a ghost
    for (int l = 0; l < input.nb_critical; l++) {
      int e_l = tau - 1 - input.kmers_frequency[input.list_criticals[l]];
      lhs = 0;
      for (int i = 0; i < nb_context; i++) {
        for (int j = 0; j < input.alphabet.size(); j++) {
          lhs += input.alpha[l][i][j] * x[i][j];
        }
      }
      lhs -= z[l] * k * input.nb_hashmark;
      model.addConstr(lhs <= e_l, "limit occurences or count as ghost");
    }

    // All hashmark must have a replacement
    for (int i = 0; i < nb_context; i++) {
      lhs = 0;
      for (int j = 0; j < input.alphabet.size(); j++) {
        lhs += x[i][j];
      }
      model.addConstr(lhs == input.context_hashmark_count[i],
                      "all hashmark are replaced");
    }

    // Optimize model
    model.optimize();
    int status = model.get(GRB_IntAttr_Status);

    if ((status == GRB_INF_OR_UNBD) || (status == GRB_INFEASIBLE) ||
        (status == GRB_UNBOUNDED)) {
      cout << "The model cannot be solved "
           << "because it is infeasible or unbounded" << endl;
      return 1;
    }
    if (status != GRB_OPTIMAL) {
      cout << "Optimization was stopped with status " << status << endl;
      return 1;
    }

    // Print x and z result
    /*for (int i = 0; i < nb_context; i++) {
      for (int j = 0; j < input.alphabet.size(); j++) {
        cout << "x[" << i << "][" << j << "]: " << x[i][j].get(GRB_DoubleAttr_X)
             << endl;
      }
    }*/

    int sum_z = 0;
    for (int l = 0; l < input.nb_critical; l++) {
      // cout << "z[" << l << "]: " << z[l].get(GRB_DoubleAttr_X) << endl;
      sum_z += z[l].get(GRB_DoubleAttr_X);
    }
    cout << "Number of ghost: " << sum_z << endl;

  } catch (GRBException e) {
    cout << "Error code = " << e.getErrorCode() << endl;
    cout << e.getMessage() << endl;
  } catch (...) {
    cout << "Exception during optimization" << endl;
  }

  return 0;
}

File: ilp/HM_ilp.py
#!/usr/bin/env python3.7

# Copyright 2020, Gurobi Optimization, LLC

# This example formulates and solves the following simple MIP model:
#  maximize
#        x +   y + 2 z
#  subject to
#        x + 2 y + 3 z <= 4
#        x +   y       >= 1
#        x, y, z binary

import gurobipy as gp
from gurobipy import GRB

from collections import defaultdict
import sys
import operator
from pprint import pprint

# Input: k tau input_file forbiden_patterns_file
k = int(sys.argv[1])
tau = int(sys.argv[2])
text_file = sys.argv[3]
pattern_file = sys.argv[4]
S = ""
forbiden_patterns = []

with open(text_file, "r") as file:
    S = file.read().replace("\n", "")

with open(pattern_file, "r") as file:
    forbiden_patterns = set(file.read().split("\n"))


def pos_sharp(T):
    # function returns the position of the first #
    # if there is one, else returns -1
    for j in range(len(T)):
        if T[j] == "#":
            return j
    return -1


i = 0
kmers_occ = defaultdict(int)
occ_context_hashmark = defaultdict(int)
nb_hashmark = 0
alphabet = set()
# Compute the ditribution of context in front and behind
while i + k < len(S):
    p = pos_sharp(S[i : i + k])
    if p == -1:
        # Get all the kmer of context
        while i + k - 1 < len(S) and S[i + k - 1] != "#":
            kmers_occ[S[i : i + k]] += 1
            alphabet.add(S[i + k - 1])
            i += 1
        if i + k - 1 < len(S):
            occ_context_hashmark[(S[i : i + k - 1], S[i + k : i + 2 * k - 1])] += 1
            nb_hashmark += 1
        i += k
    else:
        occ_context_hashmark[
            (S[i + p - k + 1 : i + p], S[i + p + 1 : i + p + k - 1])
        ] += 1
        nb_hashmark += 1
        i += p + 1
context_hashmark = list(occ_context_hashmark)
alphabet = list(alphabet) + [""]

forbiden_replacement = set()
critical_kmers = {}
nb_critical = 0
alpha = []

for i in range(len(context_hashmark)):
    context = context_hashmark[i]
    for j in range(len(alphabet)):
        letter = alphabet[j]
        replacement = context[0] + letter + context[1]
        for start in range(len(replacement) - k + 1):
            kmer = replacement[start : start + k]
            if kmer in forbiden_patterns:
                # kmer is forbiden
                forbiden_replacement.add((i, j))
            elif kmers_occ[kmer] < tau and kmers_occ[kmer] + k * nb_hashmark >= tau:
                # kmer is critical
                if kmer in critical_kmers:
                    alpha[critical_kmers[kmer]][i, j] += 1
                else:
                    critical_kmers[kmer] = nb_critical
                    nb_critical += 1
                    alpha.append(
                        dict(
                            (val, 0)
                            for val in [
                                (i, j)
                                for i in range(len(context_hashmark))
                                for j in range(len(alphabet))
                            ]
                        )
                    )
                    alpha[critical_kmers[kmer]][i, j] += 1

list_critical = [""] * nb_critical
for kmer, order in critical_kmers.items():
    list_critical[order] = kmer

try:

    # Create a new model
    m = gp.Model("Hide and Mine")

    # Create variables
    x = m.addVars(len(context_hashmark), len(alphabet), vtype="I", name="x")
    z = m.addVars(nb_critical, vtype="B", name="z")

    # Set objective
    m.setObjective(z.sum(), GRB.MINIMIZE)

    for i in range(len(context_hashmark)):
        for j in range(len(alphabet)):
            m.addConstr(x[i, j] >= 0)

    for i, j in forbiden_replacement:
        m.addConstr(x[i, j] == 0, "no sensitive pattern")

    for l in range(len(alpha)):
        e_l = tau - 1 - kmers_occ[list_critical[l]]
        m.addConstr(
            x.prod(alpha[l]) - z[l] * k * nb_hashmark <= e_l,
            "limit occurences or ghost",
        )

    for i in range(len(context_hashmark)):
        print(occ_context_hashmark[context_hashmark[i]], context_hashmark[i])
        m.addConstr(
            x.sum(i, "*") == occ_context_hashmark[context_hashmark[i]],
            "all hashmark are replaced",
        )
    # Add constraint: x + 2 y + 3 z <= 4
    # m.addConstr(x <= 4, "c0 ")

    # Add constraint: x + y >= 1
    # m.addConstr(x + y >= 1, "c1")

    # Optimize model
    m.optimize()

    for v in m.getVars():
        print("%s %g" % (v.varName, v.x))

    print("Obj: %g" % m.objVal)

except gp.GurobiError as e:
    print("Error code " + str(e.errno) + ": " + str(e))

except AttributeError:
    print("Encountered an attribute error")
